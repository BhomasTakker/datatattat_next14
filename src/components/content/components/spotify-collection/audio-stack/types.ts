import { SearchType } from "@/types/api/spotify";
import { CollectionItem } from "@/types/data-structures/collection/item/item";

export type SpotifyCollectionItem = CollectionItem & {
	id: string;
	media: {
		type: SearchType;
	};
};

export type AudioStackClientProps = {
	items: SpotifyCollectionItem[];
};

export type SpotifyArticleProps = {
	item: SpotifyCollectionItem;
	onClick: (index: number) => void;
	index: number;
};
