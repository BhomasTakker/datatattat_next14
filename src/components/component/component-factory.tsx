import {
	ComponentsMap,
	ComponentsOptions,
} from "@/components/component/component-map";
import { PageComponent } from "@/types/page";

export const ComponentFactory = ({
	component,
}: {
	component: PageComponent;
}) => {
	const { componentType } = component || {};

	const Component = ComponentsMap.get(componentType as ComponentsOptions);

	// do better error handling
	if (!Component) {
		return <div>Component not found</div>;
	}

	return <Component component={component} />;
};
