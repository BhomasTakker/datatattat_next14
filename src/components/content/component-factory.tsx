import {
	ComponentsMap,
	ComponentsOptions,
} from "@/components/content/component-map";
import { PageComponent } from "@/types/page";

export const ComponentFactory = ({
	component,
}: {
	component: PageComponent;
}) => {
	const { componentType } = component || {};

	const Component = ComponentsMap.get(componentType as ComponentsOptions);

	// do better error handling
	if (!Component) {
		return <div>Component not found</div>;
	}

	return <Component component={component} />;
};
