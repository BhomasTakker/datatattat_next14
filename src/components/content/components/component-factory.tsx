import { getData } from "@/actions/data/get-data";
import {
	ComponentsMap,
	ComponentsOptions,
} from "@/components/content/components/component-map";
import { PageComponent } from "@/types/page";
import { UnknownObject } from "@/types/utils";

export const ComponentFactory = async ({
	component,
}: {
	component: PageComponent;
}) => {
	const { componentType } = component || {};

	const Component = ComponentsMap.get(componentType as ComponentsOptions);

	// do better error handling
	if (!Component) {
		return <div>{`Component not found:- ${componentType}`}</div>;
	}

	const { _with: queryObject } = component;

	const data = (await getData(queryObject)) as UnknownObject;

	// all components take in a component object
	// AND and a data object
	// data object contains data and any data functions
	// data functions can be used to refecth or manipulate queries
	return <Component component={component} dataObject={{ data }} />;
};
