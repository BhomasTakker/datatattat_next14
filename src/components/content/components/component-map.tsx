import { ArticleCollection } from "./article-collection/article-collection";
import { ContentOembed } from "./content-oembed/content-oembed";
import { DisplayPlayer } from "./display-player/display-player";
import { MediaPlayer } from "./media-player/media-player";

export type ComponentType =
	| typeof ArticleCollection
	| typeof ContentOembed
	| typeof DisplayPlayer
	| typeof MediaPlayer;

export enum ComponentsOptions {
	ArticleCollection = "ArticleCollection",
	ContentOembed = "ContentOembed",
	DisplayPlayer = "DisplayPlayer",
	MediaPlayer = "MediaPlayer",
}

export const ComponentsMap = new Map<ComponentsOptions, ComponentType>([
	[ComponentsOptions.ArticleCollection, ArticleCollection],
	[ComponentsOptions.ContentOembed, ContentOembed],
	[ComponentsOptions.DisplayPlayer, DisplayPlayer],
	[ComponentsOptions.MediaPlayer, MediaPlayer],
]);
