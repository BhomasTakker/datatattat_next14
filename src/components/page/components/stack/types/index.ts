import { PageStackCollectionVariants } from "@/types/components/page/stack";

export type VerticalStackProps = {
	variant: PageStackCollectionVariants.Vertical;
};

export const ContainerWidth = {
	XXS: "180",
	XS: "240",
	SM: "320",
	MD: "640",
	LG: "768",
	XL: "1024",
	XXL: "1280",
} as const;

export const ContainerWidthOptions = {
	XXS: "XXS",
	XS: "XS",
	SM: "SM",
	MD: "MD",
	LG: "LG",
	XL: "XL",
	XXL: "XXL",
} as const;

type ContainerWidth = (typeof ContainerWidth)[keyof typeof ContainerWidth];

export type ContainerWidthOptions = keyof typeof ContainerWidth;

export const ContainerHeight = {
	XXS: "240",
	XS: "360",
	SM: "480",
	MD: "600",
	LG: "720",
	XL: "840",
	XXL: "960",
} as const;

export const ContainerHeightOptions = {
	XXS: "XXS",
	XS: "XS",
	SM: "SM",
	MD: "MD",
	LG: "LG",
	XL: "XL",
	XXL: "XXL",
} as const;

type ContainerHeight = (typeof ContainerHeight)[keyof typeof ContainerHeight];

type ContainerHeightOptions = keyof typeof ContainerHeight;
export const containerWidthOptions: ContainerWidthOptions[] = Object.keys(
	ContainerWidth
) as ContainerWidthOptions[];

// Row configuration for row-stack layout
export type Row = {
	columns: number;
	evenColumns: boolean;
	// you wouldn't max/min height - you would set a height
	maxHeight: ContainerHeightOptions;
	minWidth: ContainerWidthOptions;
	// these are all actually strings!
	// Unless we can/want to convert these after http requests use strings
	index?: string; // optional for default row - specifies which row this config applies to
};

export type Column = {
	minWidth: ContainerWidthOptions;
	maxHeight: ContainerHeightOptions;
	evenColumns: boolean;
};

export type RowStackProps = {
	variant: PageStackCollectionVariants.RowStack;
	defaultRow: Row; // Default configuration used for any row not specifically configured
	rows?: Row[]; // Optional array of row-specific configurations - index property determines which row each config applies to
};

export type PageStackProps = {
	variant: PageStackCollectionVariants;
} & (VerticalStackProps | RowStackProps);
