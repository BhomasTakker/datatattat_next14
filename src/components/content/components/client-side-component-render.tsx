"use client";

import { PageComponent } from "@/types/page";
import { UnknownObject } from "@/types/utils";
import { useEffect, useState } from "react";
import { ComponentType } from "./component-map";
import { With } from "@/types/component";

type ClientSideComponentProps = {
	Component: ComponentType;
	component: PageComponent;
	getData: (queryObject: With) => Promise<UnknownObject>;
};

export const ClientSideComponent = ({
	getData,
	Component,
	component,
}: ClientSideComponentProps) => {
	const { _with: componentQueryObject, componentProps } = component;
	const { _with: componentPropsQueryObject } = componentProps || {};
	const [componentData, setComponentData] = useState<UnknownObject | null>(
		null
	);
	const [isError, setIsError] = useState<boolean>(false);

	useEffect(() => {
		const getComponentData = async () => {
			try {
				const data = await getData(
					componentQueryObject || componentPropsQueryObject
				);
				setComponentData(data);
			} catch {
				setIsError(true);
			}
		};
		getComponentData();
	}, []);

	if (isError) {
		return <div>Error loading data</div>;
	}

	if (!componentData) {
		return <div>Loading...</div>;
	}

	return (
		<Component component={component} dataObject={{ data: componentData }} />
	);
};
