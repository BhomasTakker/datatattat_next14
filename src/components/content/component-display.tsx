import { PageComponent } from "@/types/page";
import { ComponentProfile } from "./component-profile/component-profile";
import { ComponentFactory } from "./components/component-factory";
import { ComponentProfileProps } from "@/types/component";
import styles from "./component-display.module.scss";

export const ComponentDisplay = ({
	component,
}: {
	component: PageComponent;
}) => {
	const { componentProps } = component;

	return (
		<section className={styles.component} data-testid="component">
			<ComponentProfile profile={componentProps as ComponentProfileProps} />
			<ComponentFactory component={component} />
		</section>
	);
};
