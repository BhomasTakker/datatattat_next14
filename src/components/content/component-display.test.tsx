import "@testing-library/jest-dom";
import { render, screen } from "@testing-library/react";
import { ComponentDisplay } from "./component-display";
import {} from "./components/component-factory";
import { PageComponent } from "@/types/page";

jest.mock("./component-profile/component-profile", () => {
	return {
		__esModule: true,
		ComponentProfile: () => {
			return <div data-testid="component-profile">Componet Profile</div>;
		},
	};
});

// In reality this would be an async component
// How would you test this?
jest.mock("./components/component-factory", () => {
	return {
		__esModule: true,
		ComponentFactory: ({ component }: { component: PageComponent }) => {
			const { componentType } = component;
			return <div data-testid="component-factory">{componentType}</div>;
		},
	};
});

describe("Component Display Test Suite", () => {
	const MOCK: PageComponent = {
		componentType: "Component",
		// @ts-expect-error - typing is not correct
		componentProps: {
			variant: "default",
			componentTitle: "Sample Title",
			componentTitleLink: "http://example.com",
			showComponentTitle: true,
		},
		_with: {
			type: "",
			query: {
				provider: "",
				params: {},
				conversions: {
					response: {},
					sub: {},
					conversionId: "",
				},
				queryId: "",
			},
		},
	};

	it("should render a component", () => {
		render(<ComponentDisplay component={MOCK} />);
		const component = screen.getByTestId("component");
		expect(component).toBeInTheDocument();
	});

	it("should render a component profile", () => {
		render(<ComponentDisplay component={MOCK} />);
		const component = screen.getByTestId("component-profile");
		expect(component).toBeInTheDocument();
	});

	it("should render a component factory", () => {
		render(<ComponentDisplay component={MOCK} />);
		const component = screen.getByTestId("component-factory");
		expect(component).toBeInTheDocument();
	});

	it("renders ComponentDisplay unchanged", () => {
		const { container } = render(<ComponentDisplay component={MOCK} />);
		expect(container).toMatchSnapshot();
	});
});
