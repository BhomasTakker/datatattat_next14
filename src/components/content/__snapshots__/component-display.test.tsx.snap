// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component Display Test Suite renders ComponentDisplay unchanged 1`] = `
<div>
  <section
    class="component"
    data-testid="component"
  >
    <div
      data-testid="component-profile"
    >
      Componet Profile
    </div>
    <div
      data-testid="component-factory"
    >
      Component
    </div>
  </section>
</div>
`;
